### PR TITLE
[SPARK-33398] Fix loading tree models prior to Spark 3.0

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
In https://github.com/apache/spark/pull/21632/files#diff-0fdae8a6782091746ed20ea43f77b639f9c6a5f072dd2f600fcf9a7b37db4f47, a new field `rawCount` was added into `NodeData`, which cause that a tree model trained in 2.4 can not be loaded in 3.0/3.1/master;
field `rawCount` is only used in training, and not used in `transform`/`predict`/`featureImportance`. So I just set it to -1L.

### Why are the changes needed?
to support load old tree model in 3.0/3.1/master

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
those models are created by following lines, in 2.4.7:
```

import org.apache.spark.ml.classification._
import org.apache.spark.ml.regression._


val data = spark.read.format("libsvm").load("/d0/Dev/Opensource/spark/data/mllib/sample_libsvm_data.txt")


val dtc = new DecisionTreeClassifier()
val dtcm = dtc.fit(data)
dtcm.save("/tmp/dtc-2.4.7")


val rfc = new RandomForestClassifier().setNumTrees(2)
val rfcm = rfc.fit(data)
rfcm.save("/tmp/rfc-2.4.7")


val gbtc = new GBTClassifier().setMaxIter(2)
val gbtcm = gbtc.fit(data)
gbtcm.save("/tmp/gbtc-2.4.7")




val dtr = new DecisionTreeRegressor()
val dtrm = dtr.fit(data)
dtrm.save("/tmp/dtr-2.4.7")


val rfr = new RandomForestRegressor().setNumTrees(2)
val rfrm = rfr.fit(data)
rfrm.save("/tmp/rfr-2.4.7")


val gbtr = new GBTRegressor().setMaxIter(2)
val gbtrm = gbtr.fit(data)
gbtrm.save("/tmp/gbtr-2.4.7")
```

**Comment by srowen** [comment]
So most of the large number of files are just new test files saved by 2.x? OK.
I think it's fine to fix and add tests. this should go into 3.0.x too?

**Comment by dongjoon-hyun** [comment]
Hi, @zhengruifeng .
I removed `[3.0][3.1]` from the PR title, `[SPARK-33398][3.0][3.1] Fix loading tree models prior to Spark 3.0`.
`[3.1]` is used when the PR is opened for `branch-3.1`. Please don't use it like an `Affected Version` from the JIRA issue~ Thanks. :)

**Comment by zhengruifeng** [comment]
@srowen Most of those files are the model files, I will try to reduce the number of files.

> this should go into 3.0.x too?

Yes,  in 3.0.1, I can not load any tree model created in 2.4.7

**Comment by zhengruifeng** [comment]
Tree model will create datafrme with numPartitions=`defaultParallelism`, which will generate lots of small files; while other models like `LinearSVC` will repartition(1) before saving.
It maybe nice to reduce the num of files in another ticket.

**Comment by zhengruifeng** [comment]
also ping @imatiach-msft

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala [hunk 1]
```diff
@@ -401,8 +403,13 @@ private[ml] object DecisionTreeModelReadWrite {
     }
 
     val dataPath = new Path(path, "data").toString
-    val data = sparkSession.read.parquet(dataPath).as[NodeData]
-    buildTreeFromNodes(data.collect(), impurityType)
+    var df = sparkSession.read.parquet(dataPath)
```
Comment (bmarcott) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: I guess spark schema doesn't allow setting default values (and converted from case class) to enable something like below?
spark.read.schema(nodeDataSchema).parquet(dataPath).as[NodeData]

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: CC @zhengruifeng on this question

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: I think it is possible to use `spark.read.schema` here, but need to list all the fields manually (method `schema` needs a `schema: StructType` or `schemaString: String` as input).
So I think it maybe more complicated.

Comment (bmarcott) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: you can create the StructType automatically from the case class like:
`Encoders.product[NodeData].schema`
...but I couldn't find a way to utilize default values well, so the new column is all null.

**Comment by srowen** [comment]
Merged to master/3.1/3.0

**Comment by zhengruifeng** [comment]
Thanks @srowen @bmarcott  for reviewing!

**Comment by WeichenXu123** [comment]
Good works!

**Comment by bmarcott** [comment]
@zhengruifeng Thanks for putting in the fix!

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala (hunk 1 above)
Comment (Vishvin95) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: @bmarcott Was this change applied for Spark 3.0.1 as well? I am facing the same error in Spark 3.0.1 libs. And I checked the above code inside the imported libraries and could see that the change has not been applied there.

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: It was reported vs 3.0.1, see JIRA. Looks like 3.0.2

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala:406]: https://github.com/apache/spark/pull/30889#issuecomment-749872161

see may previous comments, and this fix come in 3.0.2

**Comment by purple-dude** [comment]
Hi All,
I have trained a random forest model in pyspark version 2.4 but I am unable to reload it in pyspark version 3.0.3 but it gives below error :
spark3_load_model_error

Please suggest how should I proceed ?

**Pull request closed without merging**